cmd/kubeadm/app/cmd/upgrade/apply.go

```diff
@@ -0,0 +1,172 @@
+	cmd.Flags().BoolVar(&flags.force, "force", flags.force, "Force upgrading although some requirements might not be met")
+	cmd.Flags().BoolVar(&flags.dryRun, "dry-run", flags.dryRun, "Do not change any state, just output what actions would be applied")
+	cmd.Flags().StringVar(&flags.newK8sVersionStr, "version", flags.newK8sVersionStr, "Selected version to upgrade to")
+
+	return cmd
+}
+
+// RunApply takes care of the actual upgrade functionality
+func RunApply(flags *applyFlags) error {
+
+	// Start with the basics, verify that the cluster is healthy and get config
+	client, cfg, err := EnforceRequirements(flags.parent.kubeConfigPath, flags.parent.cfgPath, flags.parent.printConfig)
+	if err != nil {
+		return err
+	}
+
+	// Set the version on the config object now
+	cfg.KubernetesVersion = flags.newK8sVersionStr
+
+	internalcfg := &kubeadmapi.MasterConfiguration{}
+	api.Scheme.Convert(cfg, internalcfg, nil)
+
+	fmt.Printf("[upgrade/version] You have chosen to upgrade to version %q\n", flags.newK8sVersionStr)
+	versionGetterImpl := upgrade.NewKubeVersionGetter(client, os.Stdout)
+	versionSkewErrs := upgrade.EnforceVersionPolicies(versionGetterImpl, flags.newK8sVersionStr, flags.newK8sVersion, flags.parent.allowExperimentalUpgrades, flags.parent.allowRCUpgrades)
+	if versionSkewErrs != nil {
+
+		if len(versionSkewErrs.Mandatory) > 0 {
+			fmt.Printf("[upgrade/version] FATAL: The --version argument is invalid: %v.\n[upgrade/version] Can be bypassed if you pass the --force flag\n", versionSkewErrs.Mandatory)
+			return fmt.Errorf("The --version argument is invalid")
+		}
+
+		if len(versionSkewErrs.Skippable) > 0 {
+			fmt.Printf("[upgrade/version] ERROR: The --version argument is invalid: %v.\n[upgrade/version] Can be bypassed if you pass the --force flag\n", versionSkewErrs.Skippable)
+
+			if !flags.force {
+				return fmt.Errorf("The --version argument is invalid")
+			}
+			fmt.Println("[upgrade/version] Continuing since the --force flag is set")
+		}
+	}
+
+	if !flags.flagYes && !flags.dryRun {
+
+		fmt.Print("[upgrade/confirm] Are you sure you want to proceed with the upgrade? [y/N]: ")
+
+		scanner := bufio.NewScanner(os.Stdin)
+		scanner.Scan()
+		if err = scanner.Err(); err != nil {
+			return fmt.Errorf("couldn't read from standard input: %v", err)
+		}
+		answer := scanner.Text()
+		if strings.ToLower(answer) != "y" {
+			return fmt.Errorf("won't proceed; the user didn't answer (Y|y) in order to continue")
+		}
+	}
+
+	// Use a prepuller implementation based on creating DaemonSets
+	prepuller := upgrade.NewDaemonSetPrepuller(client, internalcfg)
+	// Block until all DaemonSets are ready
+	// TODO: Make this timeout configurable?
+	upgrade.PrepullImagesInParallell(prepuller, 15*time.Minute)
+
+	if upgrade.IsControlPlaneSelfHosted(client) {
+		fmt.Printf("[upgrade/apply] Upgrading your Self-Hosted control plane to version %q...\n", flags.newK8sVersionStr)
+
+		// Upgrade a self-hosted cluster
+		return fmt.Errorf("not implemented")
```

Comment:
maybe we should fail earlier than this for self-hosted clusters? I'd rather know straight away that's it's not yet implemented than after a few minutes of optimistic expectation